compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -2978,20 +2978,76 @@ static TR::Register * inlineSinglePrecisionSQRT(TR::Node *node, TR::CodeGenerato
   return node->getRegister();
 }
 
-TR::Register* OMR::X86::TreeEvaluator::performSimpleAtomicMemoryUpdate(TR::Node* node, TR_X86OpCodes op, TR::CodeGenerator* cg)
+static TR::Register* inlineAtomicMemoryUpdate(TR::Node* node, TR_X86OpCodes op, TR::CodeGenerator* cg)
    {
    TR_ASSERT((!TR_X86OpCode(op).hasLongSource() && !TR_X86OpCode(op).hasLongTarget()) || TR::Compiler->target.is64Bit(), "64-bit instruction not supported on IA32");
-   TR::Register* address = cg->evaluate(node->getChild(0));
-   TR::Register* value   = cg->evaluate(node->getChild(1));
-   TR::Register* result  = cg->allocateRegister();
+   auto address = cg->evaluate(node->getChild(0));
+   auto value   = cg->gprClobberEvaluate(node->getChild(1), MOVRegReg());
 
-   generateRegRegInstruction(MOVRegReg(), node, result, value, cg);
-   generateMemRegInstruction(op, node, generateX86MemoryReference(address, 0, cg), result, cg);
+   generateMemRegInstruction(op, node, generateX86MemoryReference(address, 0, cg), value, cg);
 
-   node->setRegister(result);
+   node->setRegister(value);
    cg->decReferenceCount(node->getChild(0));
    cg->decReferenceCount(node->getChild(1));
-   return result;
+   return value;
+   }
+
+static TR::Register* inlineAtomicCompareAndMemoryUpdate(TR::Node* node, TR_X86OpCodes op, bool returnVaue, TR::CodeGenerator* cg)
+   {
+   TR_ASSERT((!TR_X86OpCode(op).hasLongSource() && !TR_X86OpCode(op).hasLongTarget()) || TR::Compiler->target.is64Bit(), "64-bit instruction not supported on IA32");
+   auto address  = cg->evaluate(node->getChild(0));
+   auto oldvalue = cg->gprClobberEvaluate(node->getChild(1), MOVRegReg());
+   auto newvalue = cg->evaluate(node->getChild(2));
+
+   TR::RegisterDependencyConditions* deps = generateRegisterDependencyConditions((uint8_t)1, (uint8_t)1, cg);
+   deps->addPreCondition(oldvalue, TR::RealRegister::eax, cg);
+   deps->addPostCondition(oldvalue, TR::RealRegister::eax, cg);
+
+   generateMemRegInstruction(op, node, generateX86MemoryReference(address, 0, cg), newvalue, deps, cg);
+   if (!returnVaue)
+      {
+      generateRegInstruction(SETE1Reg, node, oldvalue, cg);
+      generateRegRegInstruction(MOVZXReg4Reg1, node, oldvalue, oldvalue, cg);
+      }
+
+   node->setRegister(oldvalue);
+   cg->decReferenceCount(node->getChild(0));
+   cg->decReferenceCount(node->getChild(1));
+   cg->decReferenceCount(node->getChild(2));
+   return oldvalue;
+   }
+
+static TR::Register* inline64BitAtomicCompareAndMemoryUpdateOn32Bit(TR::Node* node, bool returnVaue, TR::CodeGenerator* cg)
+   {
+   TR_ASSERT(TR::Compiler->target.is32Bit(), "32-bit only");
+   auto address  = cg->evaluate(node->getChild(0));
+   auto oldvalue = cg->longClobberEvaluate(node->getChild(1));
+   auto newvalue = cg->evaluate(node->getChild(2));
```

Comment:
Same nit as above with the use of `auto`.

Reply:
Sure, updated to `TR::Register*`